### ISSUE TITLE
Support PyPy2.7 and PyPy3.5 v5.7

### ISSUE THREAD

**Issue by Juanlu001** [travis-ci/travis-ci#7507]
PyPy2.7 and PyPy3.5 v5.7 have just been released:

https://morepypy.blogspot.com.es/2017/03/pypy27-and-pypy35-v57-two-in-one-release.html

And PyPy3.5 v5.7 is the first PyPy3 to support Python 3.5, thanks to the Mozilla grant.

I'm aware there are at least other two issues tracking PyPy support on Travis CI, namely https://github.com/travis-ci/travis-ci/issues/6727 and https://github.com/travis-ci/travis-ci/issues/6277. However it's not clear what's the status of any of those.

**Comment by webknjaz** [comment]
@Juanlu001 You may try hacking it like I did: https://github.com/open-craft-guild/aiohttp-ripozo/blob/4f1c963b59387a3637fa31fda4500d5d604a9520/.travis.yml#L10-L61

**Comment by BanzaiMan** [comment]
We rely on [php-build](https://github.com/pyenv/pyenv/tree/master/plugins/python-build/share/python-build) to guide our archiving process. We would need these releases to be supported by it before we can build the archive and provide support for it.

Looking at [the current files](https://github.com/pyenv/pyenv/tree/01d39850b564876a624fe9352b25ce69f4f59f0f/plugins/python-build/share/python-build), `pypy-5.7.1` is used for pypy2.7 v5.7.1 (and this is [available](https://travis-ci.org/BanzaiMan/travis_production_test/jobs/239821950#L99) now), but pypy3.5 is not. I suggest opening an issue (or, even better, a PR) to add the pypy3.5 v5.7.1 definition.

**Comment by Juanlu001** [comment]
> We rely on php-build to guide our archiving process.

You mean python-build, right? Right?? 😅

> Looking at the current files, pypy-5.7.1 is used for pypy2.7 v5.7.1 (and this is available now), but pypy3.5 is not. I suggest opening an issue (or, even better, a PR) to add the pypy3.5 v5.7.1 definition.

Is this what you mean?

https://github.com/pyenv/pyenv/blob/master/plugins/python-build/share/python-build/pypy3.5-5.7.1-beta

https://github.com/squeaky-pl/portable-pypy#latest-python-35-release

**Comment by BanzaiMan** [comment]
Yes. I mean `python-build`. 😆

**Comment by BanzaiMan** [comment]
@Juanlu001 That is a definition for a beta release, and it is not suitable for our purposes.

**Comment by ionelmc** [comment]
Why does it matter. It's opt-in anyway.

**Comment by BanzaiMan** [comment]
@ionelmc It matters because the deviant naming scheme costs development and testing time when we develop tooling around it.

`pypy` works now. See https://github.com/travis-ci/travis-build/pull/1073.

**Comment by BanzaiMan** [comment]
@amcgregor We can look at PyPy "nightly" options, based on [`pypy-dev`](https://github.com/pyenv/pyenv/blob/master/plugins/python-build/share/python-build/pypy-dev) and [`pypy3-dev`](https://github.com/pyenv/pyenv/blob/master/plugins/python-build/share/python-build/pypy3-dev). Please open a separate issue.

**Comment by ionelmc** [comment]
I don't really get it (I'm not familiar with the build system either). Are you saying the extra "-beta" part is a problem?

Bottom line is that pypy3 5.7.1 (regardless if some people think it's beta quality or not) is the best pypy3 yet 😁

**Comment by BanzaiMan** [comment]
@ionelmc
> I don't really get it (I'm not familiar with the build system either). Are you saying the extra "-beta" part is a problem?

Yes. That's what I mean.

As I said before, if you want the best PyPy ever, please open an issue/PR with yyuu/pyenv. Thanks!

**Comment by Juanlu001** [comment]
@ionelmc If PyPy3 5.5 was alpha and 5.7 is beta, perhaps it is safe to assume we're approaching an stable release? I'm not familiar with the PyPy release process though.

@BanzaiMan Thank you for your response - If you're encouraging us to open a new issue for nightly PyPy releases (which I'm willing to) does this mean this can be closed as PyPy2 already works? And while we are at it, would you please review the status of #6727?

**Comment by amcgregor** [comment]
@BanzaiMan Will do; apologies.  Submitted as #7878.

**Comment by BanzaiMan** [comment]
Speaking of naming, `pypy` is not exactly the best name going forward. python-build currently uses `pypy2-*` and `pypy3-*`; ideally, we should be using `pypy2-x.y`, `pypy3-x.y` as aliases for `pypy2-x.y.z` and `pypy3-x.y.z`. (and drop `pypy`, `pypy3`, etc.)

**Comment by BanzaiMan** [comment]
https://github.com/travis-ci/travis-ci/issues/6727#issuecomment-307186234

These runtimes are available for on-demand installation:

For Precise
```
$ aws s3 ls s3://travis-python-archives/binaries/ubuntu/12.04/x86_64/ | grep pypy.*bz2$
2015-11-02 11:23:02   24721563 pypy-2.6.1.tar.bz2
2015-12-20 21:39:16   25425728 pypy-4.0.1.tar.bz2
2015-12-20 21:39:16   25425728 pypy-4.0.tar.bz2
2016-07-08 01:14:27   28990441 pypy-5.0.0.tar.bz2
2016-07-08 01:23:58   28920422 pypy-5.0.1.tar.bz2
2016-07-08 01:21:35   29039673 pypy-5.1.1.tar.bz2
2016-07-08 01:21:49   29094467 pypy-5.1.tar.bz2
2016-07-08 01:18:24   33100456 pypy-5.3.1.tar.bz2
2016-07-08 01:22:47   33113882 pypy-5.3.tar.bz2
2016-11-09 17:29:25   33614598 pypy-5.4.1.tar.bz2
2016-09-08 04:55:35   33571606 pypy-5.4.tar.bz2
2017-06-05 23:57:16   29295218 pypy-5.6.0.tar.bz2
2017-06-05 23:57:16   29295218 pypy-5.6.tar.bz2
2016-09-07 22:10:25   35466496 pypy3-2.4.0.tar.bz2
2015-12-20 21:42:38   35064187 pypy3-2.4.tar.bz2
2016-07-12 15:47:52   31092855 pypy3.3-5.2-alpha1.tar.bz2
2017-03-08 23:09:08   26550720 pypy3.3-5.5-alpha.tar.bz2
```

For Trusty

```
$ aws s3 ls s3://travis-python-archives/binaries/ubuntu/14.04/x86_64/ | grep pypy.*bz2$
2015-11-02 11:23:12   27706390 pypy-2.6.1.tar.bz2
2015-12-20 21:39:09   25482627 pypy-4.0.1.tar.bz2
2015-12-20 21:39:09   25482627 pypy-4.0.tar.bz2
2016-07-08 01:12:53   28912705 pypy-5.0.0.tar.bz2
2016-07-08 01:24:09   28923467 pypy-5.0.1.tar.bz2
2016-07-08 01:21:36   29015127 pypy-5.1.1.tar.bz2
2016-07-08 01:25:50   29111536 pypy-5.1.tar.bz2
2016-07-08 01:27:40   33130244 pypy-5.3.1.tar.bz2
2016-07-08 01:23:30   33149149 pypy-5.3.tar.bz2
2016-11-09 17:30:26   33625737 pypy-5.4.1.tar.bz2
2016-09-08 04:55:55   33515769 pypy-5.4.tar.bz2
2017-06-05 23:57:09   29302524 pypy-5.6.0.tar.bz2
2017-06-05 23:57:08   29302524 pypy-5.6.tar.bz2
2017-06-05 23:57:46   29048662 pypy-5.7.1.tar.bz2
2017-06-05 23:57:46   29048662 pypy-5.7.tar.bz2
2016-09-07 22:10:32   42801682 pypy3-2.4.0.tar.bz2
2015-12-20 21:42:15   41986352 pypy3-2.4.tar.bz2
2016-07-12 15:48:47   31161218 pypy3.3-5.2-alpha1.tar.bz2
2017-03-08 23:09:11   26665269 pypy3.3-5.5-alpha.tar.bz2
```

pypy3.5-5.7 will be compiled when it is added to python-build.

**Comment by ionelmc** [comment]
@BanzaiMan 

>  > I don't really get it (I'm not familiar with the build system either). Are you saying the extra "-beta" part is a problem?
> 
> Yes. That's what I mean.

So if pypy3.3-5.5-alpha is available why can't pypy3-5.7.1-beta be available? Hard to keep straight face here :)

**Comment by BanzaiMan** [comment]
See https://github.com/pyenv/pyenv/issues/931.

**Comment by BanzaiMan** [comment]
> So if pypy3.3-5.5-alpha is available why can't pypy3-5.7.1-beta be available? Hard to keep straight face here :)

Because it the alpha was manually built. I am working on tooling, so that we don't have to do it by hand ever again.

**Comment by BanzaiMan** [comment]
pypy3.5-5.7.1-beta is now available for Trusty. https://travis-ci.org/BanzaiMan/travis_production_test/builds/241402192#L97

It is not possible to support this version on Precise. See https://github.com/pyenv/pyenv/issues/925#issuecomment-307527973 for more details.

I believe this can now be closed.
